tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,344 @@
+// <copyright file="TraceUtil.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.TraceProcessors
+{
+    internal class TraceUtil
+    {
+        // MaxTagLength the maximum length a tag can have
+        private const int MaxTagLength = 200;
+
+        private static readonly UTF8Encoding Encoding = new UTF8Encoding(false);
+
+        // https://github.com/DataDog/datadog-agent/blob/0454961e636342c9fbab9e561e6346ae804679a9/pkg/trace/traceutil/truncate.go#L34-L49
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TruncateUTF8(ref string value, int limit)
+        {
+            if (string.IsNullOrEmpty(value) || Encoding.GetByteCount(value) <= limit)
+            {
+                return false;
+            }
+
+            var valueCharArray = value.ToCharArray();
+            for (var i = 0; i < valueCharArray.Length; i++)
+            {
+                if (Encoding.GetByteCount(valueCharArray, 0, i) > limit)
+                {
+                    value = value.Substring(0, i - 1);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/0454961e636342c9fbab9e561e6346ae804679a9/pkg/trace/agent/normalizer.go#L183-L188
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsValidStatusCode(string statusCode)
+        {
+            if (int.TryParse(statusCode, out int code))
+            {
+                return 100 <= code && code < 600;
+            }
+
+            return false;
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/0454961e636342c9fbab9e561e6346ae804679a9/pkg/trace/traceutil/normalize.go#L93-L204
+        public static string NormalizeTag(string value)
```

Comment:
That was fun to read and was my annual reminder that C# has a `goto` statement 😅 If the unit tests pass then it's all good to me 👍🏼